Fixes #30507

timsuchanek opened the pull request:
<!--
Thank you for submitting a pull request!

Please verify that:
* [ ] There is an associated issue in the `Backlog` milestone (**required**)
* [ ] Code is up-to-date with the `master` branch
* [ ] You've successfully run `gulp runtests` locally
* [ ] There are new or updated unit tests validating the change

Refer to CONTRIBUTING.MD for more details.
  https://github.com/Microsoft/TypeScript/blob/master/CONTRIBUTING.md
-->

Fixes https://github.com/microsoft/TypeScript/issues/30507

I saw, that when the object expression has incorrect properties, that don't exist on the type, the autocompletion works fine.
The decision, if the constraint type should be taken as the default or not is being made in [checker.ts:15733](https://github.com/microsoft/TypeScript/compare/master...timsuchanek:genericObjectExpressionCompletion?expand=1#diff-c3ed224e4daa84352f7f1abcd23e8ccaR15733) and as this is about 10 function calls deep when coming from https://github.com/microsoft/TypeScript/blob/master/src/services/completions.ts#L1457 I had to introduce a parameter in all these functions that can be passed down.
I saw, that there was already a `ContextFlags` type, which can be passed in to `getContextualType`, so I extended it and to make it accessible from the `completions.ts`, I had to add it to the `TSChecker` types in https://github.com/microsoft/TypeScript/compare/master...timsuchanek:genericObjectExpressionCompletion?expand=1#diff-4b8bd1eea29904f1be39cd864e1a45c0R3335

Let me know if I could have achieved this in a different way, but after many hours of debugging, this looked like the "most minimal" change to achieve the goal.

I added 2 tests, all tests are green.

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/microsoft/TypeScript?pullRequest=32100) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/TypeScript?pullRequest=32100) before we can accept your contribution.<br/><br/>:x: timsuchanek [sign now](https://cla.opensource.microsoft.com/microsoft/TypeScript?pullRequest=32100)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/microsoft/TypeScript?pullRequest=32100) it.</sub>

schickling commented:
Very excited to see this fix! Thanks for your work on this @timsuchanek!

timsuchanek commented:
@typescript-bot run dt

timsuchanek commented:
Sorry to bother, but maybe @ahejlsberg you can have a look as you introduced the `ContextFlags`. Not sure if I'm "misusing" them.

AviVahl commented:
Does it also fix #28470?

timsuchanek commented:
@AviVahl from my understanding yes! I'll fix the new conflicts tomorrow, then it's again ready to get merged.

timsuchanek commented:
@AviVahl I added a fix so that your use-case with the constructor arguments is also supported and also added the 2 examples from https://github.com/microsoft/TypeScript/issues/28470 as test cases.

@weswigham @ahejlsberg can you prioritize this PR so that it doesn't get stale again? Thanks 🙏

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -16023,7 +16018,8 @@ namespace ts {
                 if (constraint) {
                     const instantiatedConstraint = instantiateType(constraint, context.nonFixingMapper);
-                    if (!inferredType || !context.compareTypes(inferredType, getTypeWithThisArgument(instantiatedConstraint, inferredType))) {
+                    const isCompletionContext = contextFlags && ((contextFlags | ContextFlags.Completion) === contextFlags);
```
Why not just `contextFlags & ContextFlags.Completion`?

timsuchanek replied:
That's very true! It still has to be `contextFlags && (contextFlags & ContextFlags.Completion)` as `contextFlags` can be undefined.
Just updated it.

DanielRosenwasser commented:
@timsuchanek would you also be able to add a test for string literal completions too?

```ts
declare function get<T, key: K extends keyof T>(obj: T, key: K): T[K];

get({ hello: 123, world: 456 }, "/**/");
```

This is one of the most-maddening and longest-standing issues we've had in completions for a long time. It would be really great to get this reviewed @ahejlsberg @weswigham @andrewbranch.

timsuchanek commented:
@DanielRosenwasser I just simplified the "binary and" statement and also added your example as a test.

Important to note: Your example also worked before my PR, I added it to the tests nonetheless.

andrewbranch commented on src/compiler/checker.ts:
```diff
@@ -16097,7 +16092,8 @@ namespace ts {
                     const instantiatedConstraint = instantiateType(constraint, context.nonFixingMapper);
-                    if (!inferredType || !context.compareTypes(inferredType, getTypeWithThisArgument(instantiatedConstraint, inferredType))) {
+                    const isCompletionContext = contextFlags && (contextFlags & ContextFlags.Completion);
+                    if (!inferredType || !context.compareTypes(inferredType, getTypeWithThisArgument(instantiatedConstraint, inferredType)) || isCompletionContext) {
```
Nit: put `isCompletionContext` before `!context.compareTypes(...)` for more efficient short-circuiting for completions

timsuchanek replied:
Good point! Just fixed it.

timsuchanek commented:
Thanks @andrewbranch! Is there a way that I can accept the baselines?
Locally on my machine the tests succeed, can't reproduce the baseline failure.

weswigham commented:
Pass `--skipPercent=0`, and, failing that, set the `CI` environment variable.

weswigham commented:
(also rebase/merge with master)

andrewbranch commented:
I fixed the baseline merge. @weswigham @ahejlsberg any reservations about exporting `ContextFlags` in the public API?

schickling commented:
Thanks a lot @andrewbranch. Can’t wait to see this merged! (This is a big deal for us at @prisma.) 🎉

sheetalkamat commented on tests/baselines/reference/api/typescript.d.ts:
```diff
@@ -2026,6 +2026,11 @@ declare namespace ts {
          */
         runWithCancellationToken<T>(token: CancellationToken, cb: (checker: TypeChecker) => T): T;
     }
+    export enum ContextFlags {
```
I think this as well as new signature need to be internal

timsuchanek replied:
Do you mean we should use something different than the `ContextFlags` when calling the code from the completion service to prevent this being type being exposed?

andrewbranch replied:
I think Sheetal means to move `ContextFlags` into one of the `ts` namespaces that’s already marked `/** @internal */` and then to mark the second parameter of `TypeChecker['getContextualType']` as `/* @internal */` (or declare it as an internal overload, but the linter might call that an unnecessary overload)

andrewbranch replied:
Just filed #33350. I think it would be ok to [ignore the line from the linter](https://palantir.github.io/tslint/usage/rule-flags/). @sheetalkamat?

sheetalkamat replied:
Yes in past we have marked the overload internal and disabled the rule for that line.

timsuchanek replied:
Done. [Had to adjust the tests](https://github.com/microsoft/TypeScript/pull/32100/commits/96a782ee4c426e58d0a2ad9f00f85e86f78c010c) because of https://github.com/microsoft/TypeScript/pull/32266

Anything else you need from my side to get this merged?

andrewbranch replied:
Cool, this doesn’t even touch the public API baselines now. Would be good to get an approval from another person on the team but I think this is good.

weswigham commented on src/compiler/checker.ts:
```diff
@@ -16201,7 +16196,8 @@ namespace ts {
                     const instantiatedConstraint = instantiateType(constraint, context.nonFixingMapper);
-                    if (!inferredType || !context.compareTypes(inferredType, getTypeWithThisArgument(instantiatedConstraint, inferredType))) {
+                    const isCompletionContext = contextFlags && (contextFlags & ContextFlags.Completion);
+                    if (!inferredType || isCompletionContext || !context.compareTypes(inferredType, getTypeWithThisArgument(instantiatedConstraint, inferredType))) {
```
This doesn't seem like the right place to do this - rather than passing the flag all the way thru signature resolution, if this flag is set, we should just look at the "base signature" (`getBaseSignature`) rather than starting inference at all, I think. Then again, if we did that we wouldn't get an instantiated constraint which reports more detailed information on circular constraints. Then again again, I'm pretty sure as-is this is caching into the `resolvedSignature` of a node, which this very much is not - which means requesting completions poisons the cache with incorrect types for later checking.

timsuchanek replied:
Thanks for the feedback @weswigham! Do you have a suggestion how we still could fix the autocompletion for this? This was, unfortunately, the only way I found (with my limited knowledge of the codebase).

Maybe introducing a new method as an alternative to `getContextualType` just to be used by the completion service?

weswigham replied:
I think `getContextualType` is still right, we just need to avoid using any signature caches.

orta commented:
@typescript-bot pack this

andrewbranch commented:
Hey @timsuchanek, I took a stab at addressing @weswigham’s concerns with some guidance from @DanielRosenwasser. I didn’t want to push a more substantial change here so I branched off of this PR and pushed a new one up at #33937. Would greatly appreciate your input there. Thanks!

andrewbranch commented:
@timsuchanek, thanks for championing this through! Your commits here have been merged into master via #33937. 🏆

timsuchanek commented:
Thanks a lot, @andrewbranch for pushing this over the finish line!
